File: dmopc/dmpg17g2/dmpg17g2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class dmpg17g2 {
	public static void main(String[] args) throws Exception {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		String[] tokens = in.readLine().split(" ");
		int n = Integer.parseInt(tokens[0]);
		int q = Integer.parseInt(tokens[1]);
		int[] swords = new int[n];
		tokens = in.readLine().split(" ");
		for (int i = 0; i < n; i++) {
			swords[i] = Integer.parseInt(tokens[i]);
		}
		for (int i = 0; i < q; i++) {
			tokens = in.readLine().split(" ");
			if (tokens[0].equals("S")) {
				swords[Integer.parseInt(tokens[1]) - 1] = Integer.parseInt(tokens[2]);
			} else if (tokens[0].equals("Q")) {
				long have = 0;
				long high = -1000000000;
				for (int j = Integer.parseInt(tokens[1]) - 1; j < Integer.parseInt(tokens[2]); j++) {
					have += swords[j];
					high = have > high ? have : high;
					have = have < 0 ? 0 : have;
				}
				System.out.println(high);
			}
		}
		in.close();
	}
}

File: dmopc/dmpg17g2/dmpg17g2.py
input = __import__("sys").stdin.readline
n, q = map(int, input().split())
swords = map(int, input().split())
for _ in xrange(q):
	q = input().split()
	if q[0] == "S":
		swords[int(q[1]) - 1] = int(q[2])
	elif q[0] == "Q":
		have = 0
		high = -1000000000
		for i in xrange(int(q[1]) - 1, int(q[2])):
			have += swords[i]
			high = max(high, have)
			have = max(have, 0)
		print(high)